### DISCUSSION TITLE
[Intel] Overclocking bins with unlocked K processors [Experimental]

### DISCUSSION THREAD

**Discussion by cyring** [cyring/CoreFreq#492] [Show and tell]
Commit db15a83fb87856e095a3925f8b83205cd2eef1d5 displays the frequency ratio bins of **unlocked** processors. Usually the **K series**.

* Frequency and ratio bins are presented as relative values
  Those are read from [`MSR_FLEX_RATIO`](https://github.com/cyring/CoreFreq/blob/cf26a206d5330a9d3f899962748892798329296d/x86_64/intel_reg.h#L669)
  
![2024-06-01-224100_644x340_scrot](https://github.com/cyring/CoreFreq/assets/11563789/3f80c8a7-e875-4560-bbe8-c3cc1c3e7e37)

* `OC` capability is shown in `Technologies`
  `OC_ENABLED` bit(`29`) from `CAPID_B` CSR register at offset `0xE8`

![2024-06-01-224025_644x340_scrot](https://github.com/cyring/CoreFreq/assets/11563789/85240027-19d0-4f55-aaac-ef5a079a851a)

----

A non K processor will display as below

* Uncapable OC

![2024-06-01-225054_644x340_scrot](https://github.com/cyring/CoreFreq/assets/11563789/4c01579a-c357-4d41-9898-8c43e0c4b307)

* No `Overclock` section within `Processor` window

![2024-06-01-225107_644x340_scrot](https://github.com/cyring/CoreFreq/assets/11563789/7b474525-8765-43aa-8cae-b1b83402e140)

**Comment by cyring** [comment]
@rushvora @olejon @BugReporterZ @justanerd and anyone with an unlocked Intel processor

Based on [`develop`](https://github.com/cyring/CoreFreq/tree/develop) branch, can you please show me the results of the above ?

Remark: because MSR FLEX is poorly documented, you have to activate the experimental mode while driver is starting 
```
insmod corefreqk.ko Experimental=1
```

Any processor with successful access will be marked for feature production in next commits

**Reply by BugReporterZ** [reply]
This is what I'm seeing:

![image](https://github.com/cyring/CoreFreq/assets/26941368/8abb3f07-65a8-4818-be2e-3395867e380a)
![image](https://github.com/cyring/CoreFreq/assets/26941368/eb11e9f3-efe3-4694-9b35-19bc3be8d24b)
![image](https://github.com/cyring/CoreFreq/assets/26941368/62e03e62-cdba-4cd5-a2fc-d5c757be7f33)

**Reply by cyring** [reply]
Thanks a lot, I will add this processor to the granted list.

In case you have other Intel architectures, can you show me results ?

**Comment by cyring** [comment]
Some legacy Xeon appear to have a OC ratio within `MSR_FLEX_RATIO` but at different bits field named `CLOCK_FLEX_MAX[12-8]`

Right after booting, the `MSR_FLEX_RATIO` with my Xeon W3690 is zeroed but I noticed than I can write a certain bits pattern, thus `CLOCK_FLEX_MAX`

```sh
wrmsr 0x194 0x10400
rdmsr -a 0x194
10400
10400
10400
10400
10400
10400
10400
10400
10400
10400
10400
10400
```

Here is the result: 

![2024-06-02-162625_642x546_scrot](https://github.com/cyring/CoreFreq/assets/11563789/cb372500-8eae-4115-81ff-5d4971983931)

The function has been improved with two methods:
One made for legacy architectures; the second for _recent_ processors.

@BugReporterZ Can you please tell if you are still reading the same results since commit f622261545aec2f3d9ef441a7e9959109f664a1c

**Reply by BugReporterZ** [reply]
After doing:
```
git pull
make clean all MAX_FREQ_HZ=5555000000
sudo insmod ./build/corefreqk.ko Experimental=1
sudo ./build/corefreqd 
```

I see:
![image](https://github.com/cyring/CoreFreq/assets/26941368/fa0d33c5-864a-4320-8e9d-b64a49758ea1)
![image](https://github.com/cyring/CoreFreq/assets/26941368/7926f463-08de-4c8e-b7a3-e60799a403fb)
![image](https://github.com/cyring/CoreFreq/assets/26941368/b9b05271-d61d-4502-a61d-c31c629b60f3)


BTW: I need to add `MAX_FREQ_HZ=5555000000` when building, otherwise the top frequency displayed in the main view will be the maximum reported HWP frequency of 6400 MHz (it's been like this for quite a while) instead of the actual top frequency of my CPU (currently 5000 MHz).

![image](https://github.com/cyring/CoreFreq/assets/26941368/a91af004-76e0-47c2-b841-a37616995325)

**Reply by cyring** [reply]
Thanks for the non regression tests

About the top ratio in the ruler, it is conceived as the max value of all ratios, including those from HWP.

I wouldn't touch `MAX_FREQ_HZ` which is set `7125000000` in Makefile and serves as a max scaling value anong various functions.

Can't you set the max HWP to your needs ? 
Or disable HWP in BIOS in favor of your original P-State frequencies ?

**Reply by cyring** [reply]
Edit:
I had changed to `7125000000` because it was the maximum HWP capability found between Alder and Raptor cpuid bits

**Reply by BugReporterZ** [reply]
> Can't you set the max HWP to your needs ? 

If I decrease the HWP frequency, it decreases all CPU frequencies proportionally to the change, it turns out.

> Or disable HWP in BIOS in favor of your original P-State frequencies ?

HWP / Intel Speed Shift has much lower latencies than the older SpeedStep algorithms, so I would prefer to keep it enabled.

**Reply by cyring** [reply]
> > Can't you set the max HWP to your needs ?
> 
> If I decrease the HWP frequency, it decreases all CPU frequencies proportionally to the change, it turns out.

HWP-Max is at most the same value as Turbo 1C in generations prior Alder Lake. See [Wiki](https://github.com/cyring/CoreFreq/wiki/CPU-support) for references.

I don't know why Intel changed that. Perhaps to please with hybrid architectures.

But UI ruler and its top value have been designed long time a go and I don't want to break that feature.

Also CPPC with AMD Zen is still aligning HWP Max with the Core Performance Boost ratio.

I'm waiting for  CoreFreq outputs of Meteor Lake to see if it's becoming the architectural trend.

**Reply by cyring** [reply]
@BugReporterZ I'm able to reproduce your issue in a similar way with Turbo ratios rather than HWP ratios

In BIOS, when Ivybridge `3770K` **non turbo ratio** is overclocked, the other turbo ratios are set to `63` which are wrongly decoded by _CoreFreq_ as meaningful frequency ratios

I'm still searching for discriminant bits to consider or not those ratios in the aggregation of the max ratio.

https://github.com/cyring/CoreFreq/blob/f622261545aec2f3d9ef441a7e9959109f664a1c/x86_64/corefreq-cli.c#L166

**Reply by BugReporterZ** [reply]
I should point out that the reported HWP max frequency changes depending on the CPU ratio configuration. In the other screenshot I had a very slightly different P-core ratio configuration (50, 50, 49, 49, 48, 48, 47, 47) and it showed 64, with with a bone stock one (50, **49**, 49, 49, 48, 48, 47, 47), the max is now 63.

![image](https://github.com/cyring/CoreFreq/assets/26941368/ce69c419-fb0b-46fa-8a02-4aea64794780)

**Reply by cyring** [reply]
Can you please try this one ?
The UI should display the ratios on ruler accordingly to HWP and Turbo features states

[CoreFreq_develop.tar.gz](https://github.com/user-attachments/files/15589858/CoreFreq_develop.tar.gz)

**Reply by BugReporterZ** [reply]
I don't see any obvious change.

![image](https://github.com/cyring/CoreFreq/assets/26941368/4a7e4b63-4ffb-477b-a3db-a734d82899ce)

**Reply by cyring** [reply]
Does it make a difference when switching off Turbo in `Technologies` window ?

**Reply by BugReporterZ** [reply]
I tried disabling `Turbo Boost Max 3.0` and this happened:

![image](https://github.com/cyring/CoreFreq/assets/26941368/756d18d1-4760-4ade-bad6-642c26432ab1)

Load frequencies seemed unaffected, here during a rendering load at 125W (PL1):

![image](https://github.com/cyring/CoreFreq/assets/26941368/affae475-2c9a-43cf-a2e6-f37c633cfc4a)

**Reply by cyring** [reply]
Do you have removed `MAX_FREQ_HZ` from your build ?

**Reply by BugReporterZ** [reply]
I used the default, which appears to be 7125 MHz or so. In other words, I didn't explicitly set that.

![image](https://github.com/cyring/CoreFreq/assets/26941368/94bdce23-aaa7-49fa-9eca-24789321d57c)

**Reply by cyring** [reply]
Can you give a try to this one:
[CoreFreq_develop.tar.gz](https://github.com/user-attachments/files/15593208/CoreFreq_develop.tar.gz)

Also can you set `HWP-MAX` equal `Turbo 1C` ?
Here hardware only allows me lower HWP to Turbo and not the opposite

![2024-06-05-175304_644x354_scrot](https://github.com/cyring/CoreFreq/assets/11563789/a1e0d865-d46f-4102-bc54-b79e9f0eb6da)

**Reply by BugReporterZ** [reply]
These are my HWP-Max settings:
![image](https://github.com/cyring/CoreFreq/assets/26941368/74518318-d057-423f-8312-951f671991c3)

Tried to changed them like this:
![image](https://github.com/cyring/CoreFreq/assets/26941368/34e15c07-4fa8-4c91-8816-1c3c8706b350)

Single core won't exceed 3900 MHz:
![image](https://github.com/cyring/CoreFreq/assets/26941368/7c12f997-09e0-4126-9b7b-c9f182451aa0)

All-core also shows the same max value:
![image](https://github.com/cyring/CoreFreq/assets/26941368/fe9705af-57ee-4fdd-9169-1c000f1ec247)

However the voltage seems too high for this frequency. I tried an all-core rendering benchmark and performance appeared to be worse than usual at 125W, so it only decreased frequencies without decreasing voltages as well, meaning it's now running inefficiently.

**Reply by cyring** [reply]
Is the single core frequency capped to 3.9 GHz even with the Energy Policy `HWP EPP` set to the minimum: a zero value ?

![2024-06-05-181509_644x354_scrot](https://github.com/cyring/CoreFreq/assets/11563789/3630cb89-f903-432f-81c0-3d11c7ded927)

**Reply by BugReporterZ** [reply]
Yes, same. HWP EPP only affects how quickly the maximum frequencies are reached during light loads. When set to 0, if C1E is enabled, they are only either minimum (800 MHz) or maximum (varies depending on the number of active cores).

**Reply by cyring** [reply]
> Yes, same. HWP EPP only affects how quickly the maximum frequencies are reached during light loads. When set to 0, if C1E is enabled, they are only either minimum (800 MHz) or maximum (varies depending on the number of active cores).

I see but I never face that HWP `reduction` frequency case beside `EPP`

And setting the `HWP-TGT` to 5 GHz does not make a difference neither ?

**Reply by BugReporterZ** [reply]
If I set HWP-TGT to 50 instead of AUTO, I get again a maximum frequency of 3900 MHz.

![image](https://github.com/cyring/CoreFreq/assets/26941368/6c0dafde-083e-422e-a938-b1191c0506fb)

With AUTO:

![image](https://github.com/cyring/CoreFreq/assets/26941368/a21b0335-79a8-4a18-a2ed-ce0b48690998)

**Reply by cyring** [reply]
Also may you have a better control on the HWP if _CoreFreq_ is registered as the kernel [CPU-Idle and CPU-Freq drivers](https://github.com/cyring/CoreFreq/wiki/CoreFreq-as-the-Clock-Source,-CPU-Freq-and-CPU-Idle-driver) ?

![2024-06-05-183208_635x68_scrot](https://github.com/cyring/CoreFreq/assets/11563789/cb021117-cad5-49bb-a8ba-c58fb8e1a286)

**Reply by cyring** [reply]
CPU-Idle CPU-Freq registrations are made in `Settings`

![2024-06-05-183957_268x358_scrot](https://github.com/cyring/CoreFreq/assets/11563789/5543193c-4b73-4921-80bd-0ac5320eaa04)

**Reply by BugReporterZ** [reply]
> [CPU-Idle and CPU-Freq drivers](https://github.com/cyring/CoreFreq/wiki/CoreFreq-as-the-Clock-Source,-CPU-Freq-and-CPU-Idle-driver) ?

Didn't work for me. I did as follows:

```
sudo insmod ./build/corefreqk.ko Register_ClockSource=1 Register_CPU_Freq=1 Register_Governor=1 Register_CPU_Idle=1
su
echo "corefreq_tsc" > /sys/devices/system/clocksource/clocksource0/current_clocksource
exit
sudo ./build/corefreqd
```

When I try to enable that:

![image](https://github.com/cyring/CoreFreq/assets/26941368/d900dfe4-ed2a-4577-87f5-d9bc029c183f)

**Reply by cyring** [reply]
The CPU-Idle and CPU-Freq kernel modules must not be loaded.
I suggest to blacklist them in the boot command line.
Then _CoreFreq_ is able to take their place.

**Reply by BugReporterZ** [reply]
I'm sorry, but I don't have time to do all of that and the necessary testing. For what it's worth, `turbostat` is also reporting a HWP-Max ratio of 63x. 

```
# turbostat

turbostat version 2024.04.08 - Len Brown <lenb@kernel.org>
Kernel command line: BOOT_IMAGE=/boot/vmlinuz-6.9.3-1-default root=UUID=c671c5b3-1948-4ab3-a91a-e7dedd8932e4 splash=silent quiet security=apparmor mitigations=auto
CPUID(0): GenuineIntel 0x20 CPUID levels
CPUID(1): family:model:stepping 0x6:97:2 (6:151:2) microcode 0x35
CPUID(0x80000000): max_extended_levels: 0x80000008
CPUID(1): SSE3 MONITOR SMX EIST TM2 TSC MSR ACPI-TM HT TM
CPUID(6): APERF, TURBO, DTS, PTM, HWP, HWPnotify, HWPwindow, HWPepp, HWPpkg, EPB
cpu0: MSR_IA32_MISC_ENABLE: 0x00840089 (TCC No-EIST MWAIT PREFETCH TURBO)
CPUID(7): No-SGX Hybrid
CPUID(0x15): eax_crystal: 2 ebx_tsc: 188 ecx_crystal_hz: 38400000
TSC: 3609 MHz (38400000 Hz * 188 / 2 / 1000000)
CPUID(0x16): base_mhz: 3600 max_mhz: 3800 bus_mhz: 100
cpu0: MSR_PLATFORM_INFO: 0x8040838f1812400
8 * 100.0 = 800.0 MHz max efficiency frequency
36 * 100.0 = 3600.0 MHz base frequency
cpu0: MSR_TURBO_RATIO_LIMIT: 0x2f2f303031313132
47 * 100.0 = 4700.0 MHz max turbo 8 active cores
47 * 100.0 = 4700.0 MHz max turbo 7 active cores
48 * 100.0 = 4800.0 MHz max turbo 6 active cores
48 * 100.0 = 4800.0 MHz max turbo 5 active cores
49 * 100.0 = 4900.0 MHz max turbo 4 active cores
49 * 100.0 = 4900.0 MHz max turbo 3 active cores
49 * 100.0 = 4900.0 MHz max turbo 2 active cores
50 * 100.0 = 5000.0 MHz max turbo 1 active cores
cpu0: MSR_SECONDARY_TURBO_RATIO_LIMIT: 0x2424242424242626
36 * 100.0 = 3600.0 MHz max turbo 8 active cores
36 * 100.0 = 3600.0 MHz max turbo 7 active cores
36 * 100.0 = 3600.0 MHz max turbo 6 active cores
36 * 100.0 = 3600.0 MHz max turbo 5 active cores
36 * 100.0 = 3600.0 MHz max turbo 4 active cores
36 * 100.0 = 3600.0 MHz max turbo 3 active cores
38 * 100.0 = 3800.0 MHz max turbo 2 active cores
38 * 100.0 = 3800.0 MHz max turbo 1 active cores
cpu0: MSR_CONFIG_TDP_NOMINAL: 0x00000024 (base_ratio=36)
cpu0: MSR_CONFIG_TDP_LEVEL_1: 0x00000000 ()
cpu0: MSR_CONFIG_TDP_LEVEL_2: 0x00000000 ()
cpu0: MSR_CONFIG_TDP_CONTROL: 0x80000000 ( lock=1)
cpu0: MSR_TURBO_ACTIVATION_RATIO: 0x00000000 (MAX_NON_TURBO_RATIO=0 lock=0)
cpu0: cpufreq driver: intel_pstate
cpu0: cpufreq governor: powersave
cpufreq intel_pstate no_turbo: 0
cpu0: MSR_MISC_PWR_MGMT: 0x000030c2 (ENable-EIST_Coordination ENable-EPB DISable-OOB)
cpu0: MSR_PM_ENABLE: 0x00000001 (HWP)
cpu0: MSR_HWP_CAPABILITIES: 0x010d2e3f (high 63 guar 46 eff 13 low 1)
cpu0: MSR_HWP_REQUEST: 0x80003f0b (min 11 max 63 des 0 epp 0x80 window 0x0 pkg 0x0)
cpu0: MSR_HWP_REQUEST_PKG: 0x8000ff01 (min 1 max 255 des 0 epp 0x80 window 0x0)
cpu0: MSR_HWP_INTERRUPT: 0x00000001 (EN_Guaranteed_Perf_Change, Dis_Excursion_Min)
cpu0: MSR_HWP_STATUS: 0x00000004 (No-Guaranteed_Perf_Change, Excursion_Min)
cpu0: EPB: 6 (balanced)
cpu0: MSR_IA32_POWER_CTL: 0x00fc005b (C1E auto-promotion: ENabled)
cpu0: MSR_PKG_CST_CONFIG_CONTROL: 0x74008008 (UNdemote-C1, demote-C1, locked, pkg-cstate-limit=8 (unlimited))
/dev/cpu_dma_latency: 16000 usec (constrained)
current_driver: intel_idle
current_governor: menu
current_governor_ro: menu
cpu0: POLL: CPUIDLE CORE POLL IDLE
cpu0: C1E: MWAIT 0x01
cpu0: C6: MWAIT 0x20
cpu0: C8: MWAIT 0x40
cpu0: C10: MWAIT 0x60
cpu0: MSR_PKGC3_IRTL: 0x00000000 (NOTvalid, 0 ns)
cpu0: MSR_PKGC6_IRTL: 0x00000000 (NOTvalid, 0 ns)
cpu0: MSR_PKGC8_IRTL: 0x00000000 (NOTvalid, 0 ns)
cpu0: MSR_PKGC10_IRTL: 0x00000000 (NOTvalid, 0 ns)
Uncore Frequency package0 die0: 800 - 4600 MHz (800 - 4600 MHz) 800 MHz
RAPL: 2097 sec. Joule Counter Range, at 125 Watts
cpu0: MSR_RAPL_POWER_UNIT: 0x000a0e03 (0.125000 Watts, 0.000061 Joules, 0.000977 sec.)
cpu0: MSR_PKG_POWER_INFO: 0x000003e8 (125 W TDP, RAPL 0 - 0 W, 0.000000 sec.)
cpu0: MSR_PKG_POWER_LIMIT: 0x4284e0001583e8 (UNlocked)
cpu0: PKG Limit #1: ENabled (125.000 Watts, 1.000000 sec, clamp ENabled)
cpu0: PKG Limit #2: ENabled (156.000 Watts, 0.002441* sec, clamp DISabled)
cpu0: MSR_VR_CURRENT_CONFIG: 0x00000000
cpu0: PKG Limit #4: 0.000000 Watts (UNlocked)
cpu0: MSR_DRAM_POWER_LIMIT: 0x00000000 (UNlocked)
cpu0: DRAM Limit: DISabled (0.000 Watts, 0.000977 sec, clamp DISabled)
cpu0: MSR_PP0_POLICY: 0
cpu0: MSR_PP0_POWER_LIMIT: 0x00000000 (UNlocked)
cpu0: Cores Limit: DISabled (0.000 Watts, 0.000977 sec, clamp DISabled)
cpu0: MSR_PP1_POLICY: 0
cpu0: MSR_PP1_POWER_LIMIT: 0x00000000 (UNlocked)
cpu0: GFX Limit: DISabled (0.000 Watts, 0.000977 sec, clamp DISabled)
cpu0: MSR_IA32_TEMPERATURE_TARGET: 0x00641400 (100 C) (100 default - 0 offset)
cpu0: MSR_IA32_PACKAGE_THERM_STATUS: 0x88450800 (31 C)
cpu0: MSR_IA32_PACKAGE_THERM_INTERRUPT: 0x02000003 (100 C, 100 C)
cpu0: MSR_MISC_FEATURE_CONTROL: 0x00000000 (L2-Prefetch L2-Prefetch-pair L1-Prefetch L1-IP-Prefetch)
```

It might have to do with the ratio between the maximum non-Turbo Boost 3.0 P-core frequency (49x) and maximum E-core frequency (38x) = 49/38 = approx 1.29

```
...
cpu0: MSR_HWP_CAPABILITIES: 0x010d2e3f (high 63 guar 46 eff 13 low 1)
cpu0: MSR_HWP_REQUEST: 0x80003f0b (min 11 max 63 des 0 epp 0x80 window 0x0 pkg 0x0)
...
```

High = 63 / 1.29 = approx. 49x
Guaranteed = 46 / 1.29 = approx. 36x (the "base frequency")
Minimum = 11 / 1.29 = approx 9x (that would be close to the idle frequency)

**Reply by cyring** [reply]
Or when some features are deactivated, like Turbo v3, Processor can be governed by common denominator: the E-core.

If I have a chance to program such hybrid processor, I may find some clues

**Reply by cyring** [reply]
Fyi changes commited in 4bc38d696fa739d7fcd7a8f542b8069a93bcbbc4

**Reply by BugReporterZ** [reply]
I see this after rebuilding:

![image](https://github.com/cyring/CoreFreq/assets/26941368/11cb9abd-5485-4457-80e4-127dd124b7bd)

**Reply by cyring** [reply]
I have reverted the changes in commit 332dbdac52918ff5c3ca17ce315fd7ef92a40726 because too many UI impacts and it was not solving for your issue.

**Reply by cyring** [reply]
@BugReporterZ Hello

In commit 9b644fd5097e7895ac9d92750984b923be071762 you can now build with the constraints `UI_RULER_MINIMUM` and `UI_RULER_MAXIMUM`

* For example:
```
make -j UI_RULER_MAXIMUM=50 UI_RULER_MINIMUM=5
```

**Reply by BugReporterZ** [reply]
It appears to work—for the screenshot below I defined a maximum of 55 and it seems to be respecting that:
![image](https://github.com/cyring/CoreFreq/assets/26941368/19110fe5-6fa7-4a4a-a37a-42bd0b269566)

**Reply by cyring** [reply]
Thank you for your return

**Comment by cyring** [comment]
@paulzzh Hello,

Can you please test the OC Bin with your `i7-14700K` ?

You'll have to run _CoreFreq_ in Experimental mode
For ex. (letter case matter)
```
insmod build/corefreqk.ko Experimental=1
```
Mode is recalled in UI/Settings 

You can then make a screenshot of the CLI [Processor] window with the OC Bin, if visible.

Thank you

**Reply by paulzzh** [reply]
![image](https://github.com/user-attachments/assets/edc5044f-06f4-49a7-8200-c4ee4130083d)

**Reply by cyring** [reply]
Great. Access is now safely granted since commit 74b8354e0312859d911c291bd45aaa3b1e3cf8bd

**Comment by cyring** [comment]
Hello,

@Technologicat Can you please test the OC Bin ?

You'll have to run _CoreFreq_ in Experimental mode

For ex. (_letter case matter_)
```
insmod build/corefreqk.ko Experimental=1
```
Mode is recalled in UI/Settings 

You can then make a screenshot of the CLI [Processor] window with the OC Bin (visible or not).

Thank you

**Comment by Technologicat** [comment]
Done, using the latest `develop`.

UI/Settings confirms Experimental ON.

Here's the screenshot of the processor window:

![image](https://github.com/user-attachments/assets/1d4ba2af-bfe8-450c-b55e-b996e7fadc8d)

**Reply by cyring** [reply]
Thank you

**Comment by cyring** [comment]
Dev ceased
